src/main/java/net/minecraftforge/common/data/SoundsProvider.java

```diff
@@ -0,0 +1,164 @@
+import net.minecraft.data.IDataProvider;
+import org.apache.commons.lang3.text.translate.JavaUnicodeEscaper;
+
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+@SuppressWarnings("deprecation")
+public abstract class SoundsProvider implements IDataProvider {
+    private final Gson GSON = (new GsonBuilder()).setPrettyPrinting().disableHtmlEscaping().create();
+    private final Map<String, Sounds> data = new TreeMap<>();
+    private final DataGenerator gen;
+    private final String modid;
+
+    public SoundsProvider(DataGenerator gen, String modid) {
+        this.gen = gen;
+        this.modid = modid;
+    }
+
+    protected abstract void addSounds();
+
+    @Override
+    public void act(DirectoryCache cache) throws IOException {
+        addSounds();
+        if (!data.isEmpty())
+            save(cache, data, this.gen.getOutputFolder().resolve("assets/" + modid + "/sounds.json"));
+    }
+
+    @Override
+    public String getName() {
+        return "Sounds.json";
+    }
+
+    private void save(DirectoryCache cache, Object object, Path target) throws IOException {
+        String data = GSON.toJson(object);
+        System.out.println(data);
+        data = JavaUnicodeEscaper.outsideOf(0, 0x7f).translate(data); // Escape unicode after the fact so that it's not double escaped by GSON
+        String hash = IDataProvider.HASH_FUNCTION.hashUnencodedChars(data).toString();
+        if (!Objects.equals(cache.getPreviousHash(target), hash) || !Files.exists(target)) {
+            Files.createDirectories(target.getParent());
+
+            try (BufferedWriter bufferedwriter = Files.newBufferedWriter(target)) {
+                bufferedwriter.write(data);
+            }
+        }
+
+        cache.recordHash(target, hash);
+    }
+
+    public void addSound(String name ,String subtitle, SoundsFile... sounds) {
+        Sounds sound = new Sounds();
+        sound.setSubtitle(subtitle);
+        sound.setSoundsFile(sounds);
+        add(name, sound);
+    }
+
+
+    public void add(String key, Sounds value) {
+        if (data.put(key, value) != null)
+            throw new IllegalStateException("Duplicate key " + key);
+    }
+
+    public class Sounds{
+
+        private SoundsFile[] sounds;
+        private String subtitle;
+
+        public Sounds(){
+
+
+        }
+
+        public void setSoundsFile(SoundsFile... soundsFile) {
+            this.sounds = soundsFile;
+        }
+
+        public void setSubtitle(String subtitle) {
+            this.subtitle = subtitle;
+        }
+
+    }
+
+    public static class SoundsFile{
+
+        public String name;
+        public Float volume;
+        public Float pitch;
+        public Integer weight;
+        public Boolean stream;
+        public Integer attenuation_distance;
+        public Boolean preload;
```

Comment:
Why are you using boxed types here? There's no way this serializes correctly.

Reply:
I use this so that when the person has not set one of the fields with a function it will not serialize so that it is not too strong

Reply:
That would still serialize the default if it was explicitly set. Better to just check that it's not the default.

Reply:
Yes but only concerns is that for example if the player wants to set the field preload to false and I check that this is different from false ...

Reply:
Just set them to the defaults of if they are not there?? For [example](https://github.com/mekanism/Mekanism/blob/1.15x/src/datagen/main/java/mekanism/client/sound/SoundBuilder.java#L116-L152). You also don't validate the dev is setting things to valid numbers, so I believe setting attenuation distance to negative will cause the json you create to crash the game/print out errors.

Reply:
And if I made an adapter type to check the fields before serializing them, it would work?